### PR TITLE
lib/protocol: Require at least 3.125% savings from compression

### PR DESCRIPTION
**Pull request by greatroar**
The new lz4 library doesn't need its output buffer to be the maximum size, unlike the old one (which would allocate if it weren't). It can take a buffer that is of a smaller size and will report if compressed data can fit inside the buffer (with a small chance of reporting a false negative). This patch uses that property by requiring compressed data to be at most n-n>>5 = .96875*n bytes long for n input bytes.

### CODE REVIEW

**Comment by greatroar** [comment]
Btw., I did some benchmarks on data collected from my running Syncthing instances and found that the new lz4 is 6x faster at compressing (1.6GB/s vs. 288MB/s) and 11x faster at decompressing (>10GB/s) even without this change, so it might be worthwhile to turn compression on by default for new folders.

**Comment by AudriusButkevicius** [comment]
We should also have a look at zstd

File: lib/protocol/protocol.go [hunk 1]
```diff
@@ -804,13 +804,13 @@ func (c *rawConnection) writeCompressedMessage(msg message, marshaled []byte, ov
 	}
 
 	cOverhead := 2 + hdrSize + 4
-	maxCompressed := cOverhead + lz4.CompressBlockBound(len(marshaled))
+	maxCompressed := cOverhead + len(marshaled) - len(marshaled)/32
```
Comment (calmh) [inline lib/protocol/protocol.go:807]: Comment to address the reason for the somewhat magical subtraction and division by 32 etc here I think would be nice

Comment (greatroar) [reply lib/protocol/protocol.go:810]: Done. Also added a tiny additional optimization.

**Review by calmh — APPROVED** [review]

**Pull request merged**